File: src/leetcode/2026/01/Solution414.java
public class Solution414 {
    public int thirdMax(int[] nums) {
        Integer max1 = null;
        Integer max2 = null;
        Integer max3 = null;

        for (Integer num : nums) {
            if (num.equals(max1) || num.equals(max2) || num.equals(max3)) {
                continue;
            }
            if (max1 == null || num > max1) {
                max3 = max2;
                max2 = max1;
                max1 = num;
            } else if (max2 == null || num > max2) {
                max3 = max2;
                max2 = num;
            } else if (max3 == null || num > max3) {
                max3 = num;
            }
        }

        return max3 != null ? max3 : max1;
    }

    public static void main(String[] args) {
        Solution414 solution = new Solution414();

        int[] nums1 = { 3, 2, 1 };
        System.out.println("Third maximum in [3, 2, 1]: " + solution.thirdMax(nums1));

        int[] nums2 = { 1, 2 };
        System.out.println("Third maximum in [1, 2]: " + solution.thirdMax(nums2));

        int[] nums3 = { 2, 2, 3, 1 };
        System.out.println("Third maximum in [2, 2, 3, 1]: " + solution.thirdMax(nums3));

        int[] nums4 = { 1, 1, 1 };
        System.out.println("Third maximum in [1, 1, 1]: " + solution.thirdMax(nums4));

        int[] nums5 = { 5, 4, 3, 2, 1 };
        System.out.println("Third maximum in [5, 4, 3, 2, 1]: " + solution.thirdMax(nums5));
    }
}

File: src/leetcode/2026/01/Solution414.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution414 {
public:
    int thirdMax(vector<int>& nums) {
        long firstMax = LONG_MIN, secondMax = LONG_MIN, thirdMax = LONG_MIN;
        for (int num : nums) {
            if (num > firstMax) {
                thirdMax = secondMax;
                secondMax = firstMax;
                firstMax = num;
            } else if (num > secondMax && num < firstMax) {
                thirdMax = secondMax;
                secondMax = num;
            } else if (num > thirdMax && num < secondMax) {
                thirdMax = num;
            }
        }
        return thirdMax == LONG_MIN ? firstMax : thirdMax;
    }
};

int main() {
    Solution414 solution;

    vector<int> nums1 = {3, 2, 1};
    int result1 = solution.thirdMax(nums1); // Expected output: 1
    cout << result1 << endl;

    vector<int> nums2 = {1, 2};
    int result2 = solution.thirdMax(nums2); // Expected output: 2
    cout << result2 << endl;

    vector<int> nums3 = {2, 2, 3, 1};
    int result3 = solution.thirdMax(nums3); // Expected output: 1
    cout << result3 << endl;

    return 0;
}